### DISCUSSION TITLE
How to create a site that references the Orchard Core source code?

### DISCUSSION THREAD

**Discussion by DrewBrasher** [OrchardCMS/OrchardCore#10588] [Q&A] [answered]
What is the best way to create a site that references the Orchard Core source code instead of NuGet packages?

I know I could clone the source code and just add my web project and any themes and modules to that Orchard Core solution but that doesn't seem like the best option because I need my projects in there own separate folder and git repository.  I also may have multiple sites that use orchard and want them to all reference the same source code for orchard. 

The other way I thought of was to create my Visual Studio solution as usual, and then use the "Add Existing Project" of visual studio to add all the orchard projects but that's a lot of projects to add.

Is there any better way of doing this? 

My main reasons for wanting this so when debugging I can step through the source and if I contribute to Orchard Core I don't have to wait for it to be approved and released to use my changes.

**Comment by hishamco** [comment]
You can use `ProjectReference` in case that you want to reference particular projects, but there's another way that I didn't remember at the moment that allows you to reference entire repo if I'm not wrong

**Reply by DrewBrasher** [reply]
Thanks for pointing me in the right direction.  Here are all the steps I did in case anyone else comes looking for this:

1. Create a solution folder called `OrchardCore`
2. Right click that folder and go to `Add > Existing Project`
3. Change the file type to `.sln`
4. Browse to the Orchard Core source and select the `OrchardCore.sln` file
5. Change all of your own projects to reference the added Orchard Core projects instead of package references

**Reply by DrewBrasher** [reply]
After doing the steps I outlined above, My solution builds without errors but when I try to launch my web project I get a 404 not found error.

**Reply by JosNun** [reply]
@DrewBrasher Did you ever find a solution to your 404 issues?

**Reply by DrewBrasher** [reply]
No, I have not found a solution yet but haven't worked on it much after I started getting the 404.  The project I was using Orchard Core for got put on hold but I still plan to find a solution when I have time.

**Reply by DrewBrasher** [reply]
@JosNun I'm working on this again but still haven't got it working yet.

**Reply by johnrom** [reply]
@DrewBrasher I just ran into this issue. The secret is in the OrchardCore repo, under the example projects. The gist of it is that when Orchard is run, the `.props` and `.targets` files listed below perform actions to register each package into the referencing package's assembly definition -- basically to act as though these themes, modules, and the Orchard Core application itself are all a part of _your_ assembly. Or something like that?? Without this, it seems like Orchard doesn't actually reference the version of MVC your application is using when it registers endpoints and things.

I think when you use the Nuget package via `<PackageReference />` it automatically includes these `.props` and `.targets`, at least that's what I'm inferring from the comments left across Orchard Core's `.csproj`'s.

For your CMS project:

https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Cms.Web/OrchardCore.Cms.Web.csproj#L3-L6

https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Cms.Web/OrchardCore.Cms.Web.csproj#L35-L39

For any userland modules referenced by your CMS project:

https://github.com/OrchardCMS/OrchardCore/blob/main/test/OrchardCore.Tests.Modules/ModuleSample/ModuleSample.csproj#L3-L4

https://github.com/OrchardCMS/OrchardCore/blob/main/test/OrchardCore.Tests.Modules/ModuleSample/ModuleSample.csproj#L14-L15

For any userland themes referenced by your CMS project:

https://github.com/OrchardCMS/OrchardCore/blob/main/test/OrchardCore.Tests.Themes/Examples.OrchardCoreThemes.Alpha/Examples.OrchardCoreThemes.Alpha.csproj#L3-L4

https://github.com/OrchardCMS/OrchardCore/blob/main/test/OrchardCore.Tests.Themes/Examples.OrchardCoreThemes.Alpha/Examples.OrchardCoreThemes.Alpha.csproj#L19-L20

So ultimately, what does actually building this into your workflow look like? Well, it looks kind of like this:

### `\Directory.Build.props` or `\Your.Project.Web\Your.Project.Web.csproj.user`

```xml
<Project>
    <!-- this probably shouldn't be checked in to Git, but let each team member decide 
        whether they want to opt into the slower but more debuggable project references -->
    <PropertyGroup>
        <LocalOrchardCoreRoot>C:\path\to\orchard\core</LocalOrchardCoreRoot>
    </PropertyGroup>
</Project>
```

### `\Your.Project.Build\OrchardCoreWebsite.props`

```xml
<Project>
  <Import Condition="'$(LocalOrchardCoreRoot)' != ''" Project="$(LocalOrchardCoreRoot)\src\OrchardCore.Build\Dependencies.props" />
  <Import Condition="'$(LocalOrchardCoreRoot)' != ''" Project="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.Application.Cms.Core.Targets\OrchardCore.Application.Cms.Core.Targets.props" />
  
  <ItemGroup Condition="'$(LocalOrchardCoreRoot)' != ''">
    <ProjectReference Include="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.Application.Cms.Targets\OrchardCore.Application.Cms.Targets.csproj" />
    <ProjectReference Include="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.Hosting.Console\OrchardCore.Hosting.Console.csproj" />
  </ItemGroup>

  <Import Condition="'$(LocalOrchardCoreRoot)' != ''" Project="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.Application.Cms.Core.Targets\OrchardCore.Application.Cms.Core.Targets.targets" />
  <Import Condition="'$(LocalOrchardCoreRoot)' != ''" Project="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.Application.Targets\OrchardCore.Application.Targets.targets" />
  <Import Condition="'$(LocalOrchardCoreRoot)' != ''" Project="$(LocalOrchardCoreRoot)\src\OrchardCore.Build\OrchardCore.Commons.targets" />
</Project>
```

### `\Your.Project.Build\OrchardCoreModule.props`

```xml
<Project>
  <Import Condition="'$(LocalOrchardCoreRoot)' != ''" Project="$(LocalOrchardCoreRoot)\src\OrchardCore.Build\Dependencies.props" />
  <Import Condition="'$(LocalOrchardCoreRoot)' != ''" Project="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.Module.Targets\OrchardCore.Module.Targets.props" />

  <ItemGroup Condition="'$(LocalOrchardCoreRoot)' == ''">    
    <PackageReference Include="OrchardCore.Module.Targets" Version="$(OrchardCoreVersion)" />
    <PackageReference Include="OrchardCore.ContentManagement" Version="$(OrchardCoreVersion)" />
    <PackageReference Include="OrchardCore.ContentTypes.Abstractions" Version="$(OrchardCoreVersion)" />
    <PackageReference Include="OrchardCore.DisplayManagement" Version="$(OrchardCoreVersion)" />    
  </ItemGroup>

  <ItemGroup Condition="'$(LocalOrchardCoreRoot)' != ''">
    <ProjectReference Include="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.Module.Targets\OrchardCore.Module.Targets.csproj" />
    <ProjectReference Include="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.ContentManagement\OrchardCore.ContentManagement.csproj" />
    <ProjectReference Include="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.ContentTypes.Abstractions\OrchardCore.ContentTypes.Abstractions.csproj" />
    <ProjectReference Include="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.DisplayManagement\OrchardCore.DisplayManagement.csproj" />
  </ItemGroup>

  <Import Condition="'$(LocalOrchardCoreRoot)' != ''" Project="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.Module.Targets\OrchardCore.Module.Targets.targets" />
</Project>
```

### `\Your.Project.Build\OrchardCoreTheme.props`

```xml
<Project>
  <Import Condition="'$(LocalOrchardCoreRoot)' != ''" Project="$(LocalOrchardCoreRoot)\src\OrchardCore.Build\Dependencies.props" />
  <Import Condition="'$(LocalOrchardCoreRoot)' != ''" Project="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.Theme.Targets\OrchardCore.Theme.Targets.props" />
  
  <ItemGroup Condition="'$(LocalOrchardCoreRoot)' == ''">
    <PackageReference Include="OrchardCore.Theme.Targets" Version="$(OrchardCoreVersion)" />
    <PackageReference Include="OrchardCore.ResourceManagement" Version="$(OrchardCoreVersion)" />
  </ItemGroup>
  
  <ItemGroup Condition="'$(LocalOrchardCoreRoot)' != ''">
    <ProjectReference Include="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.Theme.Targets\OrchardCore.Theme.Targets.csproj" />
    <ProjectReference Include="$(LocalOrchardCoreRoot)\src\OrchardCore\OrchardCore.ResourceManagement\OrchardCore.ResourceManagement.csproj" />
  </ItemGroup>
</Project>
```

### `\Your.Project.Web\Your.Project.Web.csproj`

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
  <Import Project="..\Your.Project.Build\OrchardCoreWebsite.props" />
  <!-- etc -->
</Project>
```

### `\Your.Project.Web\Your.Project.Module.csproj`, or any number of modules

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
  <Import Project="..\Your.Project.Build\OrchardCoreModule.props" />
  <!-- etc -->
</Project>
```

### `\Your.Project.Web\Your.Project.Theme.csproj`, or any number of themes

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
  <Import Project="..\Your.Project.Build\OrchardCoreTheme.props" />
  <!-- etc -->
</Project>
```

### Notes

If you only have one of each, you can obviously add these directly to your `csproj`, but I kept them separate so I can maintain them all in one place and then forget about them. 

I'm not sure if it is an issue to include the `.targets` in a `.props` file. Seems to be working fine with my current project. Your mileage may vary.

You may notice in the csproj in the OrchardCore repo, the themes also import the `Module.targets` and `Module.props`.. I don't know if that's necessary or a mistake in the projects. I removed them in mine and it seems to work OK.

**Recipes**: I am actually doing all this to try and debug recipes, and I found that the `recipe.json` from my module was not actually showing up, so I had to manually add them to the module's csproj. This has been such a rabbit hole that I have no clue if this is the correct thing to do or why it's happening, but once again it's working in my project for now.

```xml
  <ItemGroup Condition="'$(LocalOrchardCoreRoot)' != ''">
    <EmbeddedResource Include="Migrations\*.recipe.json" />
  </ItemGroup>
```

**Views**: I am not using them for this debug session, but I found that for some reason this setup doesn't copy `.Views.dll` from any userland themes or modules to `.Web\bin\*.Views.dll`. I can't figure out why, especially since it copies the Views from the Orchard modules themselves.

**Comment by DrewBrasher** [comment]
I found how to add a reference to an entire solution here:
https://docs.microsoft.com/en-us/archive/blogs/habibh/walkthrough-adding-an-existing-visual-studio-solution-to-another-solution

Here are the steps for future reference:

1. Create a solution folder called `OrchardCore`
2. Right click that folder and go to `Add > Existing Project`
3. Change the file type to `.sln`
4. Browse to the Orchard Core source and select the `OrchardCore.sln` file
5. Change all of your own projects to reference the added Orchard Core projects instead of package references

**Reply by hishamco** [reply]
What you did basically added a `ProjectReference`  to .csproj as I mentioned above ;)

**Reply by DrewBrasher** [reply]
That's right, I changed the selected answer to yours.

**Reply by hyzx86** [reply]
Hi @DrewBrasher ，
I can't find the option of * .sln here. 

>Create a solution folder called OrchardCore
>Right click that folder and go to Add > Existing Project

![image](https://user-images.githubusercontent.com/15613121/139588015-da2343a5-13a3-4e00-a0d7-92f04228e33f.png)

**Reply by DrewBrasher** [reply]
@hyzx86 I am using Visual Studio Community Edition 2019.  Is that the same as you are using?

**Reply by hyzx86** [reply]
> Visual
Yes I am using  VS2019 ，too.

**Reply by DrewBrasher** [reply]
@hyzx86 I updated to Visual Studio 2022 and now I don't see the option to add a reference to a .sln file either.

**Comment by Skrypt** [comment, marked as answer]
We are supposed to host the source code through Nuget packages symbols server. It probably doesn't work yet.
But using source code directly is also a solution for now.

**Reply by ns8482e** [reply]
Yes! Some modules has source packages published and some don’t -   Try yourself and see if you are lucky 🤞

**Reply by DrewBrasher** [reply]
@Skrypt Using the source code is what I really need because the main reason I need this is so I can go ahead use the code in https://github.com/OrchardCMS/OrchardCore/pull/10570 with my application and themes even though it has not been accepted and merged.  I've tried adding a reference to the orchard core solution in my own solution and also tried adding my projects to the orchard core solution. Both build fine but I get a 404 when I try to run my web project.

**Reply by Skrypt** [reply]
What you need to do is to reference a different version of that OrchardCore.Media module Nuget package.

You set your Orchard Core source code repository to your pull request branch.
Then you do a `dotnet pack -c Release` on this modified source code. It will create an OrchardCore.Media Nuget package locally on your PC that contains your changes.

See : https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file#packagesources

Then, on your Nuget package solution (your website); you modify the NuGet.config file so that it fetches packages standing in that folder instead of the ones hosted at Cloudsmith. I would generally just keep the OrchardCore.Media module NuGet package in that folder and remove the other ones.

You need to set the Nuget package folder feed first in the list of your Nuget feeds in that Nuget.config file so that it tries to fetch your custom modules first from that local folder.

Try this and let me know if that works.

**Reply by Skrypt** [reply]
Also, using this technique, you can use a different version of any OrchardCore module customized for your needs. 
Also, I believe that this should work when using `dotnet publish`.

Another technique is that you can simply create a custom module that references the OrchardCore.Media module and add override(s) in that module. Though, not everything can be overridden unless I'm wrong.

**Reply by DrewBrasher** [reply]
Thank you @Skrypt, creating my own NuGet packages worked!  I couldn't get it to work using just my copy of the media module but it worked when I put all the NuGet packages in my local NuGet server.

**Reply by Skrypt** [reply]
Probably related with the versioning of the Nuget packages. We use Nuget packages with versions on our Cloudsmith server. Good to hear that it worked. 😄 

See : https://github.com/OrchardCMS/OrchardCore/blob/6160a34b2bd7d20b6998b4291e7ca4bc53651a8b/.github/workflows/release_ci.yml#L67

**Reply by Skrypt** [reply]
If you want to add a contribution; you are welcome to create a small guide about how to do this as a Markdown file. This is general knowledge but that could be added somewhere in the OC Github Wiki or OC documentation.

**Reply by DrewBrasher** [reply]
I will do that.

**Reply by DrewBrasher** [reply]
I created a pull request for the documentation here: https://github.com/OrchardCMS/OrchardCore/pull/11284
This is how I did it but if you see anything that should have been done different let me know and I can update it.

**Comment by hyzx86** [comment]
I found a simple and fast way. 

Replace DLL & PDB files, and Use attach to process
https://github.com/OrchardCMS/OrchardCore/discussions/10603

**Reply by DrewBrasher** [reply]
Thanks, I will try that.

**Comment by devlife** [comment]
You could always use a git submodule to get the source code into your repo. From there it should be relatively easy to add project references as you normally would.
